### ISSUE TITLE
Asymmetric datalabel padding

### ISSUE THREAD

**Issue by drmrbrewer** [highcharts/highcharts#6213]
#### Expected behaviour

With a bounding box around the  data label, would expect uniform padding around the text.

#### Actual behaviour

More space below than above, at least in Chrome.

#### Live demo with steps to reproduce

http://jsfiddle.net/yp813csb/

Chrome (bad):
![capture](https://cloud.githubusercontent.com/assets/9009792/21711394/ac89a9a0-d3e6-11e6-8d1b-b8edfd4f9992.JPG)

IE (good):
![captureie](https://cloud.githubusercontent.com/assets/9009792/21711438/03b4ff72-d3e7-11e6-9bc5-819a88c3d35a.JPG)

I have found that enabling html improves the formatting in Chrome:

http://jsfiddle.net/yp813csb/2/

![capturechome2](https://cloud.githubusercontent.com/assets/9009792/21712046/49ce6626-d3ea-11e6-8234-19345e06cd72.JPG)

#### Affected browser(s)

See above.  At least Chrome.  And using phantomjs on the server... the same asymmetry is seen.

#### Possible solutions?

This is the same as closed issue https://github.com/highcharts/highcharts/issues/4153.  But it's still really really bugging me that I can't get something as simple as a datalabel looking good, on what is surely a very common browser (Chrome, both on desktop and mobile)... particularly if you're developing for Android!  The existing data labels just look really bad IMHO.

I tried the workaround mentioned in the above issue thread, but it got really messy.  Instead I just enable html, which as mentioned above does improve the formatting... but it leads to another problem which is that the html floats above svg as mentioned at http://www.highcharts.com/docs/chart-concepts/labels-and-string-formatting#html).

Anyway I think that a messy workaround shouldn't be necessary for such a simple component.

If it's not possible for the highcharts code to determine the spacing properly, can you at least provide separate `paddingTop`, `paddingRight`, `paddingBottom` and `paddingLeft` options for the data label, rather than just a single `padding` option as at present?  That way, the user can choose to use asymmetric padding to compensate, and thereby create a uniform padding for a particular environment?

Ideally, the separate padding options would allow you to specify a negative padding too, in case you want zero padding at top, left, right, and then you'd need negative padding at the bottom due to the extra padding there by default.

**Comment by TorsteinHonsi** [comment]
Hi, the problem is basically the same as #1101 (for IE), so I added Chrome to the same fix. See the [demo with the fix applied](http://jsfiddle.net/yp813csb/4/).

**Comment by drmrbrewer** [comment]
Is a separate fix required for rendering via phantomjs?

**Comment by TorsteinHonsi** [comment]
Apparently this doesn't happen with PhantomJS. See http://export.highcharts.com/ and Preview with the configuration object from your fiddle. 
![skjermbilde 2017-01-09 kl 10 40 40](https://cloud.githubusercontent.com/assets/227836/21762273/1a2b0454-d658-11e6-83c2-e8f5e84f0145.png)

**Comment by drmrbrewer** [comment]
OK.  I think that maybe the problem I'm still seeing when rendering with PhantomJS on the server is with specific font families, for example [TeX Gyre](http://www.gust.org.pl/projects/e-foundry/tex-gyre/) specified in the js via `chart.style.fontFamily`.  It's appearing too high in the label box:

![capture](https://cloud.githubusercontent.com/assets/9009792/21768303/dc29dbf8-d670-11e6-9e08-0b5d6685e763.JPG)

Is there any way I can move it down?  This is perhaps where individual `paddingTop`, `paddingRight`, `paddingBottom` and `paddingLeft` options for the data label would be handy, so that I could e.g. specify zero padding for top, left, and right, and a negative padding for bottom.

**Comment by TorsteinHonsi** [comment]
> Is there any way I can move it down? This is perhaps where individual paddingTop, paddingRight, paddingBottom and paddingLeft options for the data label would be handy

It would probably be handy, but it would still only be a workaround to a browser issue - also, it would be wrong in Firefox that doesn't have the same problem with bounding boxes. 

The workaround in the commit above is specific to the default font, so I think for a complete solution we would need some kind of extendable mapping for corrected bounding boxes. Something along these lines:

```js
SVGRenderer.prototype.bBoxCorrections = {
   '"Lucida Grande", "Lucida Sans Unicode", Arial, Helvetica, sans-serif': [
      ['11px', 17, 14]
   ]
}
```

.... where `11px` is the font size setting, 17 is the browser's reported box height and 14 is the corrected box height. Then you would be able to define your custom corrections like this:

```js
SVGRenderer.prototype.bBoxCorrections['TeX Gyre Adventor'] = [
   ['11px', 16, 14],
   // ... etc for other font sizes
];
```

**Comment by drmrbrewer** [comment]
Custom corrections like that would be awesome.  Would it be possible to specify a correction that was applicable to all font sizes, rather than (or as an option instead of) having to do it for all possible px values... particularly since px values are not restricted to integer values where the font size is determined programatically (but I suppose that integer px values could be enforced in the code).  For example just specifying `-10` (%) would be equivalent to `['6px', 10, 9], ['8px', 12, 10.8]` etc.

**Comment by TorsteinHonsi** [comment]
Yes a general formula would be great. It could potentially eliminate all pixel-level differences between browsers, except the glyphs themselves. I'll have to leave now, but here's something we can use for comparison: http://jsfiddle.net/highcharts/em37nvuj/

It seems like no two browsers report the same bounding boxes. Even Chrome and Safari on Mac are different.

**Comment by drmrbrewer** [comment]
Hmm, interesting... that makes it easier to see why this is so tricky.

Times on Chrome:
![chrome-times](https://cloud.githubusercontent.com/assets/9009792/21772230/4ff57afe-d682-11e6-94fb-9c125796ee9b.JPG)
Times on IE:
![ie-times](https://cloud.githubusercontent.com/assets/9009792/21772234/51e45ee8-d682-11e6-8b8b-bfcfccd037b6.JPG)

So bbox dimensions are direct from the browser, untouched by highcharts?  Chrome seems to stick to integer px heights, unlike IE.  And (maybe as a result of rounding), Chrome gives the same bbox height for 13px and 14px... both can't be right!  Or maybe it is using a different bbox height (float) to what it is reporting (rounded to int).  Not sure how on Earth one can find a pattern in there, sufficient to derive a general formula!

Can highcharts not work out for itself what the text height is?  Is this handled differently for html labels (which in my limited experience seems to render better more consistently), or does highcharts still rely on what the browser is telling it?

**Comment by TorsteinHonsi** [comment]
> So bbox dimensions are direct from the browser, untouched by highcharts?

That's correct, in this jsFiddle it comes unmodified from the DOM.

I have added pure HTML boxes [to the jsFiddle](So bbox dimensions are direct from the browser, untouched by highcharts?). The only difference here as far as I can see is that IE rounds off these numbers for the `offsetHeight`. 

> Can highcharts not work out for itself what the text height is?

We need to measure the actual box size to get label _widths_, but in theory we may be able to use logic to sort out the line _heights_ instead of reading how the browser renders it. It would be great if we could have absolute consistency between browsers. The challenge then would be to compute the bounding box height of multiline text, but these too could probably be built by adding the heights of each `tspan` that the lines are made of. I did an implementation like this in the [highcharts-jsdom experiment](https://gist.github.com/TorsteinHonsi/e8a1e6971608523eb8dd#file-highcharts-jsdom-js-L31-L82) a couple of years ago (jsdom doesn't handle bounding boxes), but it should be noted it is risky business and by no means production ready.
